utils/self-extracting-executable/decompressor.cpp

```diff
@@ -236,7 +240,16 @@ int decompressFiles(int input_fd, char * path, char * name, bool & have_compress
         files_pointer += le64toh(file_info.name_length);
         if (file_name_len != le64toh(file_info.name_length))
         {
-            strcat(file_name, ".decompressed");
+            strcat(file_name, ".decompressed.XXXXXX");
+            int fd = mkstemp(file_name);
+            if (fd == -1)
+            {
+                perror(nullptr);
+                return 1;
+            }
+            close(fd);
+            strncpy(decompressed_suffix, file_name + strlen(file_name) - 6, 6);
```

Comment:
if file was produced by the program then why fail if it exists? just rewrite it. IMHO the main concern is that some user's file with the same name exists then replacing and/or removing it will not be good - with random name we prevent it from happening
abnormal termination cleanup seems worth to implement

Reply:
>if file was produced by the program then why fail if it exists?

OK, agree, with a proper cleanup it does not matter what file it will be, with a temporary prefix or not.

BTW right code is tricky, there is global variable `decompressed_suffix` that initialized in `decompressFiles` and accessed in the caller of this function - `main`, why not fill it in main?

Reply:
...after come consideration I think it's not necessary for termination cleanup - after all it's possible there are many files to be unpacked (we support this feature though do not use it for this specific case) and cleaning up only one of them seems not very consistent... what do you think?

Reply:
>why not fill it in main?
because suffix is not just generated - function mkstemp creates actual file, and I don't want to do it prematurely

Reply:
I am for keeping things as simple as possible. If someone abnormally terminates the program (which is super unlikely), then the worst case is that some of the unpacked files are still there. Well, so be it. They don't disturb the next run and they can be deleted manually without causing harm. I fear that signal handlers will be hard to make portable across all OSs.

Reply:
>I am for keeping things as simple as possible.

I guess you are right here, also thought about this.

> If someone abnormally terminates the program (which is super unlikely)

Hit Ctrl-C because clickhouse start takes too much time, but since this compressed binary will be only in artifacts, not in packages, then I guess it is OK.

> I fear that signal handlers will be hard to make portable across all OSs.

Handling signals with `signal` is likely portable on linux/osx/freebsd.